src/liballoc/vec.rs

```diff
@@ -2128,16 +2128,34 @@ impl<T> Vec<T> {
         //      for item in iterator {
         //          self.push(item);
         //      }
-        while let Some(element) = iterator.next() {
-            let len = self.len();
-            if len == self.capacity() {
-                let (lower, _) = iterator.size_hint();
-                self.reserve(lower.saturating_add(1));
-            }
-            unsafe {
-                ptr::write(self.get_unchecked_mut(len), element);
-                // NB can't overflow since we would have had to alloc the address space
-                self.set_len(len + 1);
+        let (lower, _) = iterator.size_hint();
+        self.reserve(lower);
```

Comment:
This still has the issue of trying to reserve before advancing the iterator. taking an element and then checking the hint and capacity is strictly better because the hint will be more accurate and if you take a `None` you can skip the hint calculation and reserve attempt.

Reply:
The benefit is that it avoids calling `next` at all. While the hint should be more accurate by calling `next` it would also be even more accurate if we called `next` twice so I don't quite buy that argument. Calling `next` once would give a fast path for the empty iterator at least which seems I find a more convincing argument, however it is only a benefit if `size_hint(); reserve(0)` is slow enough to warrant this fast path.

Pushed another variant which does `next` first which should optimize better for the empty iterator. Haven't got time to benchmark it yet.

Reply:
I was more concerned about doing an suboptimal resize if we got an incorrect lower bound on the first attempt, but you're right, if it's just 0 it hopefully would be cheap enough and the next reserve will do it properly.

Reply:
Have gone a bit back and forth but I think the of reserving eagerly has a better tradeoff.

On `size_hint == 0` it does call reserve unnecessarily but it is only one more branch in an already fast case. If `size_hint > 0` then most of the time we will reserve either way and most of the time with the same value (`size_hint_minus_removed_value + 1` vs `size_hint`).

The only case where I'd expect it to be faster to call next first is if `size_hint` is expensive, despite the iterator being empty and `next` being faster to call than `size_hint` which seems unlikely (happy to be corrected however!).